blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,489 @@
+    B = rand(Float32,n,n)
+    C = rand(Float32,n,n)
+    t1 = @belapsed $A .* $B
+    t2 = @belapsed ($C .= $A .* $B)
+    t3 = @belapsed (mybroadcastavx!($C, $A, $B))
+    A,B,C = (cu(A), cu(B), cu(C))
+    t4 = @belapsed CUDA.@sync($A .* $B)
+    t5 = @belapsed CUDA.@sync($C .= $A .* $B)
+    t1,t2,t3,t4,t5
+end
+ns = 2 .^ (2:11)
+res = [alloc_timer(n) for n in ns]
+alloc      = [x[1] for x in res]
+noalloc    = [x[2] for x in res]
+noalloclv  = [x[3] for x in res]
+allocgpu   = [x[4] for x in res]
+noallocgpu = [x[5] for x in res]
+
+using Plots
+plot(ns,alloc,label="=",xscale=:log10,yscale=:log10,legend=:bottomright,
+     title="Which Micro-optimizations matter for BLAS1?",
+     ylabel = "Time (s)", xlabel = "N",
+     yticks = 10.0 .^ (-8:0.5:2),)
+plot!(ns,noalloc,label=".=")
+plot!(ns, noalloc, label="mybroadcastavx!")
+plot!(ns,allocgpu,label="= gpu")
+plot!(ns,noallocgpu,label=".= gpu")
+savefig("microopts_blas1.png")
+```
+
+![](https://user-images.githubusercontent.com/1814174/162625049-b26fd0fd-271a-4c73-a44d-f40b36b18136.png)
+
+This already highly motivates a project focused on the performance for this case, but assumptions (3) and (4) point us to additionally look at the implementation of the backpropogation. The [trade-off between different machine learning library approaches to automatic differentiation has already been discussed at length](https://www.stochasticlifestyle.com/engineering-trade-offs-in-automatic-differentiation-from-tensorflow-and-pytorch-to-jax-and-julia/), but what the general discussions can miss is the extra opportunities afforded when really specializing on a domain. Take for example the use-case inside of neural ordinary differential equations (neural ODEs) and ODE adjoints. As mentioned above, in this use case the backwards pass is applied immediately after the forward pass. Thus while [a handwritten adjoint to a neural network layer](https://github.com/SciML/DiffEqFlux.jl/blob/v1.8.1/src/fast_layers.jl#L38-L56) can look like:
+
+```julia
+ZygoteRules.@adjoint function (f::FastDense)(x,p)
+  W = ...
+  b = ...
+  r = W*x .+ b
+  y = f.σ.(r)
+  function FastDense_adjoint(v)
+    σbar = ForwardDiff.derivative.(f.σ,r)
+    zbar = v .* σbar
+    Wbar = zbar * x'
+    bbar = zbar
+    xbar = W' * zbar
+    nothing,xbar,vcat(vec(Wbar),bbar)
+  end
+  y,FastDense_adjoint
+end
+```
+
+for `sigma.(W*x .+ b)` to calculate `J'v`, you can greatly optimize this if you know that the backwards pass will immediately preceed the forward pass. Specifically, there's no need to generate closures to store values since there no indeterminate future where the gradient might be needed, instead you can immediately proceed to calculate it. And, if you're only applying it to a vector of known size `v`, then this operation can be done non-allocating by mutating a cache vector. Lastly, if we know we're only going to use the derivative w.r.t. `x` (`xbar`), then we can eliminate many calculations. Look at the simplified version:
```

Comment:
```suggestion
for `sigma.(W*x .+ b)` to calculate `J'v`, you can greatly optimize this if you know that the backwards pass will immediately preceed the forward pass. Specifically, there's no need to generate closures to store values since there no indeterminate future where the gradient might be needed, instead you can immediately proceed to calculate it. And, if you're only applying it to a vector of known size `v`, then this operation can be done without allocating by mutating a cache vector. Lastly, if we know we're only going to use the derivative w.r.t. `x` (`xbar`), then we can eliminate many calculations. Look at the simplified version:
```

SimpleChains doesn't care whether it's known size or not, it'll avoid allocating either way.
(Unless the thing of unknown size is larger than anything it's been called with previously, in which case it'll allocate [`resize!`] to increase the size of its memory buffer to be big enough.)

Reply:
That's probably worth mentioning as a side note or footnote. It's hard to put into the simple explainer though.